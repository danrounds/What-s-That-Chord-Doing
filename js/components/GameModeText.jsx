import React from 'react';
import {connect} from 'react-redux';

export class GameModeText extends React.Component {
    makeGameModeString() {
        if (this.props.inversions) {
            return {
                easyMajor: 'Easy major, with inversions',
                easyMinor: 'Easy minor. with inversions',
                intermediateMinor: 'Intermediate minor, with inversions',
                hardMajor: 'Hard major, with inversions',
                hardMinor: 'Hard minor, with inversions',
                allChords: 'All our chords, with inversions'}
            [this.props.gameType];
        } else {
            return {
                easyMajor: 'Easy major', easyMinor: 'Easy minor',
                intermediateMinor: 'Intermediate minor',
                hardMajor: 'Hard major',
                hardMinor: 'Hard minor', allChords: 'All our chords'}
            [this.props.gameType];
        }
    }

    render() {
        return (<div className="game-mode-string">{this.makeGameModeString()}</div>);
    }
}

const mapStateToProps = (state) => ({
    gameType: state.game.gameType,
    inversions: state.game.inversions
});

export default connect(mapStateToProps)(GameModeText);
