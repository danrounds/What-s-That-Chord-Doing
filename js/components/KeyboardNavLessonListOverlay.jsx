import React from 'react';
import { connect } from 'react-redux';

export class KeyboardNavLessonListOverlay extends React.Component {
    constructor() {
        super();
        this.thing = (
            <div>
              use&nbsp;
              <span className="keyHint lesson-nav-hint">w</span>&nbsp;
              <span className="keyHint lesson-nav-hint">a</span>&nbsp;
              <span className="keyHint lesson-nav-hint">s</span>&nbsp;
              <span className="keyHint lesson-nav-hint">d</span>&nbsp;
              /&nbsp;
              <span className="keyHint lesson-nav-hint">↑</span>&nbsp;
              <span className="keyHint lesson-nav-hint">←</span>&nbsp;
              <span className="keyHint lesson-nav-hint">↓</span>&nbsp;
              <span className="keyHint lesson-nav-hint">→</span>&nbsp;
              and&nbsp;
              <span className="keyHint lesson-nav-hint">Enter</span>&nbsp;
              /&nbsp;
              <span className="keyHint lesson-nav-hint">Space</span>&nbsp;
              <br/>to navigate lessons menu
            </div>);
    }

    render(props) {
        return (
            <div className="lesson-nav-keyboard-hints">
              {this.props.displayKeyboardShortcuts && this.thing}
            </div>
        );
    }
}

export default connect((state) => ({
  displayKeyboardShortcuts: state.ui.displayKeyboardShortcuts
}))(KeyboardNavLessonListOverlay);
