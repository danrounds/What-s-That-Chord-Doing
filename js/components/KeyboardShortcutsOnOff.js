import React from 'react';
import {connect} from 'react-redux';

import * as actions from  '../actions';

export class KeyboardShortcutsOnOff extends React.Component {
    constructor(props) {
        super(props);
        this.onClick = this.onClick.bind(this);
    }

    componentWillMount() {
        this.props.dispatch(actions.toggleKeyboardShortcutDisplay(
            this.props.showHints));
    }

    onClick() {
        this.props.dispatch(actions.toggleKeyboardShortcutDisplay(
            !this.props.displayKeyboardShortcuts));
    }

    render() {
        if (this.props.showButton)
            return (
                <div className="nav-button" onClick={this.onClick}>
                  {this.props.displayKeyboardShortcuts
                      ? (<p className="nav-text-keybrd">Keyboard <br/>hints <span className="nav-on">ON</span></p>)
                   : (<p className="nav-text-keybrd">Keyboard <br/>hints <span className="nav-off">OFF</span></p>)}
                </div>
            );
        else
            return null;
    }
}

const mapStateToProps = (state) => ({
    displayKeyboardShortcuts: state.game.displayKeyboardShortcuts
});

export default connect(mapStateToProps)(KeyboardShortcutsOnOff);
