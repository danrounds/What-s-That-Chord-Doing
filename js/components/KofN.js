import React from 'react';
import {connect} from 'react-redux';

export class KofN extends React.Component {
    render() {
        return (<p className="k-of-n">Question {this.props.questionNumber} of 10</p>);
    }
}

export default connect((state) => ({ questionNumber: state.game.questionNumber }))(KofN);
