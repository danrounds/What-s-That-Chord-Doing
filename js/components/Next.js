import React from 'react';
import {connect} from 'react-redux';

import * as actions from '../actions';

export class Next extends React.Component {
    constructor(props) {
        super(props);
        this.onClick = this.onClick.bind(this);
        this.state = { showKeyboardShortcuts: this.props.displayKeyboardShortcuts };
        // /\ this is kept as a "last state" variable, and compared to props
        // in the event that it changes
    }

    componentDidUpdate() {
        if (['ArrowRight', ' '].indexOf(this.props.keyValue) !== -1)
            if (this.props.questionNumber !== 10)
                this.props.dispatch(actions.getNextQuestion());
        else if (this.state.showKeyboardShortcuts !== this.props.displayKeyboardShortcuts)
            this.setState({ showKeyboardShortcuts: this.props.displayKeyboardShortcuts });
    }

    onClick() {
        if (this.props.questionNumber !== 10)
            this.props.dispatch(actions.getNextQuestion());
    }

    render() {
        return(
            <button className="nxtBtn" onClick={this.onClick}>
              next ?
              {this.props.displayKeyboardShortcuts
                  ? <div className="keyHint miniKeyHint">SPACE</div>
              : null}
            </button>
        );
    }
}

const mapStateToProps = (state, props) => ({
    keyValue: state.game.keyValue,
    displayKeyboardShortcuts: state.game.displayKeyboardShortcuts,
    questionNumber: state.game.questionNumber
});

export default connect(mapStateToProps)(Next);
