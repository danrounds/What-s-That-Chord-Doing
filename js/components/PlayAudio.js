import React from 'react';
import { connect } from 'react-redux';

export class PlayAudio extends React.Component {
    // This component plays audio--when the buttons are pressed, when we advance
    // to a new question, and when the player has answered a question correctly
    constructor(props) {
        super(props);
        this.replayPrompt = this.replayPrompt.bind(this);
        this.playIntroChordsAndPrompt = this.playIntroChordsAndPrompt.bind(this);
    }

    componentDidMount() {
        this.playSounds();
    }

    shouldComponentUpdate(nextProps) {
        return ([',','.',' '].includes(nextProps.keyPress))
            || this.props.questionNumber !== nextProps.questionNumber
            || this.props.gameNumber !== nextProps.gameNumber
            || this.props.displayKeyboardShortcuts !== nextProps.displayKeyboardShortcuts;
    }

    componentWillUpdate(nextProps) {
        if (nextProps.displayKeyboardShortcuts === this.props.displayKeyboardShortcuts)
            this.playSounds(nextProps);
    }

    playIntroChordsAndPrompt() {
        // This plays a chord progression (I-IV-V-I or i-iv-V-i) to introduce
        // our key and then plays the question prompt
        let timeOffset = 0;

        // Prompt (I-IV-V-I)
        this.props.instrument.then(piano => {
            piano.stop();
            for (let chord of this.props.introChordSequence) {
                for (let i of chord) {
                    piano.play(i, this.props.ac.currentTime + timeOffset, {duration: 0.68});
                }
                timeOffset += 0.68;
            }
            // Question chord
            this.props.instrument.then(piano => {
                for (let i of this.props.answer) {
                    piano.play(i, this.props.ac.currentTime + timeOffset, {duration: 0.68});
                }
            });
        });
    }

    replayPrompt(null_, null__, timeOffset=0) {
        // This replays the actual "question" chord. It also plays after the
        // player has made a correct guess.
        // Weird function signature is due to a quirk of our version of React,
        // in which superfluous (for our purposes) arguments are sent to
        // functions called by React with onClick. Documented: 
        // https://github.com/facebook/react/issues/8354
        this.props.instrument.then(piano => {
            piano.stop();
            for (let i of this.props.answer) {
                piano.play(i, this.props.ac.currentTime + timeOffset, {duration: 0.68});
            }
        });
    }

    playSounds(props=this.props) {
        if (!props.keyPress) {
            if (props.answeredCorrectly)
                this.replayPrompt();
            else if (!props.guessN)
                this.playIntroChordsAndPrompt();
        } else if (props.keyPress === ',') {
            this.playIntroChordsAndPrompt();
        } else if (props.keyPress === '.') {
            this.replayPrompt();
        }
    }

    render() {
        return (
            <div className="play-audio-div">
              <button className="ctrl-btn" onClick={this.playIntroChordsAndPrompt}>
                Play intro &<br/>chord again<br/>
                {this.props.displayKeyboardShortcuts
                    && <div className="keyHint miniKeyHintDiv">,</div>}
              </button>
              <button className="ctrl-btn" onClick={this.replayPrompt}>
                Play chord,<br/>again<br/>
                {this.props.displayKeyboardShortcuts
                    && <div className="keyHint miniKeyHintDiv">.</div>}
              </button>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    const processNotes = (val) => val.split('/').join('');
    let notes = state.game.notes;

    return {
        displayKeyboardShortcuts: state.ui.displayKeyboardShortcuts,
        keyPress: state.game.keyValue,
        gameNumber: state.game.gameNumber,
        answer: notes.bass ? [notes.bass, ...notes.treble].map(
            processNotes) : [],
        introChordSequence: state.game.introChordSequence.map(
            (array) => array.map(processNotes)),
        guessN: state.game.guessN,
        answeredCorrectly: state.game.answeredCorrectly,
        questionNumber: state.game.questionNumber,
    };
};

export default connect(mapStateToProps)(PlayAudio);
