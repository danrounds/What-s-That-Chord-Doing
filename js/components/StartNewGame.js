import React from 'react';
import {connect} from 'react-redux';

import * as actions from '../actions';

export class StartNewGame extends React.Component {
    constructor(props) {
        super(props);
        this.startNewGame = this.startNewGame.bind(this);
        this.showKeyboardShortcuts = this.props.displayKeyboardShortcuts;
        // /\ this is kept as a "last state" variable, and compared to props
        // in the event that it changes
    }

    componentDidUpdate() {
        if (this.props.keyValue === 'Control Enter') {
            if (this.props.gameOver || this.props.firstQ || this.prompt())
                this.props.dispatch(actions.startNewGame(this.props.gameType,
                                                         this.props.inversions));

        } else if (this.props.keyValue === 'Enter' && this.props.gameOver) {
            this.props.dispatch(actions.startNewGame(this.props.gameType,
                                                     this.props.inversions));
        } else if (this.showKeyboardShortcuts !== this.props.displayKeyboardShortcuts) {
            this.showKeyboardShortcuts = this.props.displayKeyboardShortcuts;
            this.forceUpdate();
        }
    }

    prompt() {
        return confirm('Are you sure you discard your progress?');
    }

    startNewGame() {
        if (this.props.gameOver || this.props.firstQ || this.prompt())
            this.props.dispatch(actions.startNewGame(this.props.gameType,
                                                     this.props.inversions));
    }

    render() {
        return (
            <button className="ctrlBtns" onClick={this.startNewGame}>
              Start New Game
              {this.props.displayKeyboardShortcuts
                  ? <div className="miniKeyHint"><div className="keyHint miniCtrlHint">CTRL</div><div className="miniPlus"> + </div><div className="keyHint miniEnterHint">ENTER</div></div>
              : null}
            </button>
        );
    }
}

const mapStateToProps = (state, props) => ({
    keyValue: state.game.keyValue,
    displayKeyboardShortcuts: state.game.displayKeyboardShortcuts,
    gameType: state.game.gameType,
    inversions: state.game.inversions,
    firstQ: state.game.questionNumber === 1 && !state.game.answeredCorrectly,
    gameOver: state.game.gameOver
});

export default connect(mapStateToProps)(StartNewGame);
