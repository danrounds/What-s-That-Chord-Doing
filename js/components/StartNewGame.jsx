import React from 'react';
import { connect } from 'react-redux';

import * as actions from '../actions';

export class StartNewGame extends React.Component {
    constructor() {
        super();
        this.startNewGame = this.startNewGame.bind(this);
    }

    componentDidUpdate() {
        if (this.props.keyValue === 'Control Enter') {
            if (this.props.gameOver || this.props.firstQ || this.prompt()) {
                this.props.dispatch(actions.startNewGame(this.props.gameType,
                                                         this.props.inversions));
                if (this.props.api.authToken) {
                    this.props.dispatch(actions.getMyScores(this.props.api.authToken));
                }
            }
        } else if (this.props.keyValue === 'Enter' && this.props.gameOver) {
            this.props.dispatch(actions.startNewGame(this.props.gameType,
                                                     this.props.inversions));
            if (this.props.api.authToken) {
                this.props.dispatch(actions.getMyScores(this.props.api.authToken));
            }
        }
    }

    prompt() {
        return confirm('Are you sure you discard your progress?');
    }

    startNewGame() {
        if (this.props.gameOver || this.props.firstQ || this.prompt()) {
            this.props.dispatch(actions.startNewGame(this.props.gameType,
                                                     this.props.inversions));
            if (this.props.api.authToken) {
                this.props.dispatch(actions.getMyScores(this.props.api.authToken));
            }
        }
    }

    render() {
        return (
            <button className="ctrl-btn" onClick={this.startNewGame}>
              Start new<br/>game
              {this.props.displayKeyboardShortcuts
                  && <div className="miniKeyHint"><div className="keyHint miniCtrlHint">CTRL</div><div className="miniPlus"> + </div><div className="keyHint miniEnterHint">ENTER</div></div>}
            </button>
        );
    }
}

const mapStateToProps = (state) => ({
    displayKeyboardShortcuts: state.ui.displayKeyboardShortcuts,
    api: state.api,
    keyValue: state.game.keyValue,
    gameType: state.game.gameType,
    inversions: state.game.inversions,
    firstQ: state.game.questionNumber === 1 && !state.game.answeredCorrectly,
    gameOver: state.game.gameOver,
});

export default connect(mapStateToProps)(StartNewGame);
