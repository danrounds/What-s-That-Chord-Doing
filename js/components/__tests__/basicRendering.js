import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GameModeText } from '../GameModeText';
import { KeyboardNavLessonListOverlay } from '../KeyboardNavLessonListOverlay';
import { KeyboardShortcutsOnOff } from '../KeyboardShortcutsOnOff';
import { KofN } from '../KofN';
import { LessonsNavigator } from '../LessonsNavigator';
import { Next } from '../Next';
import { PlayAudio } from '../PlayAudio';
import { StartNewGame } from '../StartNewGame';
import { Status } from '../Status';

describe('Components: Basic rendering', () => {

    test('GameModeText', () => {
        const comp = renderer.create(
            <GameModeText gameType="hardMinor" inversions={true} />
        ).toJSON();
        expect(comp).toMatchSnapshot();
    });

    test('KeyboardNavLessonListOverlay, w/ keyboard shortcuts', () => {
        const comp = renderer.create(
            <KeyboardNavLessonListOverlay displayKeyboardShortcuts={true} />
        ).toJSON();
        expect(comp).toMatchSnapshot();
    });
    test('KeyboardNavLessonListOverlay, w/out keyboard shortcuts', () => {
        const comp = renderer.create(
            <KeyboardNavLessonListOverlay displayKeyboardShortcuts={true} />
        ).toJSON();
        expect(comp).toMatchSnapshot();
    });

    test('KofN', () => {
        const comp = renderer.create(
            <KofN questionNumber={3} />
        ).toJSON();
        expect(comp).toMatchSnapshot();
    });

    test('LessonsNavigator', () => {
        const comp = mount(
            <LessonsNavigator
               lessonIndexDisplay={{ easy: true, novice: true, difficult: true, i: 1 }} />
        );
        expect(toJson(comp)).toMatchSnapshot();
    });

    test('Next, w/ keyboard shortcuts', () => {
        const comp = renderer.create(
            <Next displayKeyboardShortcuts={true} />
        ).toJSON();
        expect(comp).toMatchSnapshot();
    });
    test('Next, w/out keyboard shortcuts', () => {
        const comp = renderer.create(
            <Next displayKeyboardShortcuts={false} />
        ).toJSON();
        expect(comp).toMatchSnapshot();
    });

    const instrument = {then : () => null}, ac = {};
    test('PlayAudio, w/ keyboard shortcuts', () => {
        const comp = renderer.create(
            <PlayAudio displayKeyboardShortcuts={true} ac={ac} instrument={instrument} />
        ).toJSON();
        expect(comp).toMatchSnapshot();
    });
    test('PlayAudio, w/out keyboard shortcuts', () => {
        const comp = renderer.create(
            <PlayAudio displayKeyboardShortcuts={true} ac={ac} instrument={instrument} />
        ).toJSON();
        expect(comp).toMatchSnapshot();
    });

    test('StartNewGame, w/ keyboard shortcuts', () => {
        const comp = renderer.create(
            <StartNewGame displayKeyboardShortcuts={true} />
        ).toJSON();
        expect(comp).toMatchSnapshot();
    });
    test('StartNewGame, w/out keyboard shortcuts', () => {
        const comp = renderer.create(
            <StartNewGame displayKeyboardShortcuts={false} />
        ).toJSON();
        expect(comp).toMatchSnapshot();
    });

    test('Status, basic render w/ keyboard shortcuts', () => {
        const comp = mount(
            <Status
               displayKeyboardShorcuts={false}
               key_="Eb Major"
               chordName="G minor"
               chord="iii"
               nAnsweredRight={4}
               clicksPerRightAnswer={[3.12]}
               gameOver={true}
               />
        );
        expect(toJson(comp)).toMatchSnapshot();
    });
    test('Status, basic render, w/out keyboard shortcuts', () => {
        const comp = mount(
            <Status
               displayKeyboardShorcuts={false}
               key_="Eb Major"
               chordName="G minor"
               chord="iii"
               nAnsweredRight={4}
               clicksPerRightAnswer={[3,12]}
               gameOver={true}
               />
        );
        expect(toJson(comp)).toMatchSnapshot();
    });
    test('Status, correct answer', () => {
        const key_ = 'Eb Major', chordName = 'G minor', chord = 'iii';
        const comp = mount(
            <Status displayKeyboardShorcuts={true}
                    key_={key_}
                    chordName={chordName}
                    chord={chord}
                    answeredCorrectly={true}
                    nAnsweredRight={4}
                    clicksPerRightAnswer={[3,12]}
                    gameOver={true}
                    guessN={3}
                    />
        );
        const statusMatch = new RegExp(`[You\ got\ it\!|Yes\!|Correct\!]\ The\ `
                                       +chord+`\ chord\ of\ `+key_+`\ is\ `
                                       +chordName);
        expect(comp.find('.status-feedback').text()).toMatch(statusMatch);
        expect(comp.find('.status-secondary').at(0).text()).toMatch(/\d\ answered\ correctly/);
        expect(comp.find('.status-secondary').at(1).text()).toMatch(/\d\.\d*\ guesses\ per\ correct\ answer/);
    });
    test('Status, wrong answer', () => {
        const comp = mount(
            <Status
               displayKeyboardShorcuts={true}
               key_="Eb Major"
               chordName="G minor"
               chord="iii"
               answeredCorrectly={false}
               nAnsweredRight={4}
               guessN={4}
               clicksPerRightAnswer={[3,12]}
               gameOver={false}
               />
        );
        expect(comp.find('.status-feedback').text()).toMatch(
                /Not\ right|That\'s\ not\ it|Wrong|Incorrect|No|You\ are\ wrong|Nope/);
    });

});
