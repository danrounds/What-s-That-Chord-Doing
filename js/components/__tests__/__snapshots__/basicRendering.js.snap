// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: Basic rendering GameModeText 1`] = `
<div
  className="game-mode-string"
>
  Hard minor, with inversions
</div>
`;

exports[`Components: Basic rendering KeyboardNavLessonListOverlay, w/ keyboard shortcuts 1`] = `
<div
  className="lesson-nav-keyboard-hints"
>
  <div>
    use 
    <span
      className="keyHint lesson-nav-hint"
    >
      w
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      a
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      s
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      d
    </span>
      / 
    <span
      className="keyHint lesson-nav-hint"
    >
      ↑
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      ←
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      ↓
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      →
    </span>
      and 
    <span
      className="keyHint lesson-nav-hint"
    >
      Enter
    </span>
      / 
    <span
      className="keyHint lesson-nav-hint"
    >
      Space
    </span>
     
    <br />
    to navigate lessons menu
  </div>
</div>
`;

exports[`Components: Basic rendering KeyboardNavLessonListOverlay, w/out keyboard shortcuts 1`] = `
<div
  className="lesson-nav-keyboard-hints"
>
  <div>
    use 
    <span
      className="keyHint lesson-nav-hint"
    >
      w
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      a
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      s
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      d
    </span>
      / 
    <span
      className="keyHint lesson-nav-hint"
    >
      ↑
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      ←
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      ↓
    </span>
     
    <span
      className="keyHint lesson-nav-hint"
    >
      →
    </span>
      and 
    <span
      className="keyHint lesson-nav-hint"
    >
      Enter
    </span>
      / 
    <span
      className="keyHint lesson-nav-hint"
    >
      Space
    </span>
     
    <br />
    to navigate lessons menu
  </div>
</div>
`;

exports[`Components: Basic rendering KofN 1`] = `
<p
  className="k-of-n"
>
  Question 
  3
   of 10
</p>
`;

exports[`Components: Basic rendering LessonsNavigator 1`] = `
<LessonsNavigator
  lessonIndexDisplay={
    Object {
      "difficult": true,
      "easy": true,
      "i": 1,
      "novice": true,
    }
  }
>
  <div
    className="game lesson-nav"
    onKeyDown={[Function]}
    onMouseLeave={[Function]}
  >
    <h2
      className="modes-primary"
    >
      Game modes:
    </h2>
    <a
      className="difficultyLvl"
      href="javascript:this.onDifficultyClick"
      onClick={[Function]}
    >
      Easy
    </a>
    <div
      className="easy-container"
    >
      <ul
        className="miniLessonList"
      >
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/easy-major"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              easy major
            </a>
          </Link>
        </li>
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/easy-major-inv"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              easy major, inversions
            </a>
          </Link>
        </li>
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/easy-minor"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              easy minor
            </a>
          </Link>
        </li>
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/easy-minor-inv"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              easy minor, inversions
            </a>
          </Link>
        </li>
      </ul>
    </div>
    <a
      className="difficultyLvl"
      href="javascript:this.onDifficultyClick"
      onClick={[Function]}
    >
      Novice
    </a>
    <div
      className="novice-container"
    >
      <ul
        className="miniLessonList"
      >
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/intermediate-minor"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              intermediate minor
            </a>
          </Link>
        </li>
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/intermediate-minor-inv"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              intermediate minor, inversions
            </a>
          </Link>
        </li>
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/hard-major"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              hard major
            </a>
          </Link>
        </li>
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/hard-major-inv"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              hard major, inversions
            </a>
          </Link>
        </li>
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/hard-minor"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              hard minor
            </a>
          </Link>
        </li>
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/hard-minor-inv"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              hard minor, inversions
            </a>
          </Link>
        </li>
      </ul>
    </div>
    <a
      className="difficultyLvl"
      href="javascript:this.onDifficultyClick"
      onClick={[Function]}
    >
      Difficult
    </a>
    <div
      className="difficult-container"
    >
      <ul
        className="miniLessonList"
      >
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/all-chords"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              all chords
            </a>
          </Link>
        </li>
        <li>
          <Link
            className="nav-link"
            onMouseOver={[Function]}
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/all-chords-inv"
          >
            <a
              className="nav-link"
              onClick={[Function]}
              onMouseOver={[Function]}
              style={Object {}}
            >
              all chords, inversions
            </a>
          </Link>
        </li>
      </ul>
    </div>
  </div>
</LessonsNavigator>
`;

exports[`Components: Basic rendering Next, w/ keyboard shortcuts 1`] = `
<button
  className="ctrl-btn nxt-btn"
  onClick={[Function]}
>
  <div
    className="nxt-text"
  >
    next ?
  </div>
  <div
    className="keyHint miniKeyHint nxtKeyHint"
  >
    SPACE
  </div>
</button>
`;

exports[`Components: Basic rendering Next, w/out keyboard shortcuts 1`] = `
<button
  className="ctrl-btn nxt-btn"
  onClick={[Function]}
>
  <div
    className="nxt-text"
  >
    next ?
  </div>
</button>
`;

exports[`Components: Basic rendering PlayAudio, w/ keyboard shortcuts 1`] = `
<div
  className="play-audio-div"
>
  <button
    className="ctrl-btn"
    onClick={[Function]}
  >
    Play intro &
    <br />
    chord again
    <br />
    <div
      className="keyHint miniKeyHintDiv"
    >
      ,
    </div>
  </button>
  <button
    className="ctrl-btn"
    onClick={[Function]}
  >
    Play chord,
    <br />
    again
    <br />
    <div
      className="keyHint miniKeyHintDiv"
    >
      .
    </div>
  </button>
</div>
`;

exports[`Components: Basic rendering PlayAudio, w/out keyboard shortcuts 1`] = `
<div
  className="play-audio-div"
>
  <button
    className="ctrl-btn"
    onClick={[Function]}
  >
    Play intro &
    <br />
    chord again
    <br />
    <div
      className="keyHint miniKeyHintDiv"
    >
      ,
    </div>
  </button>
  <button
    className="ctrl-btn"
    onClick={[Function]}
  >
    Play chord,
    <br />
    again
    <br />
    <div
      className="keyHint miniKeyHintDiv"
    >
      .
    </div>
  </button>
</div>
`;

exports[`Components: Basic rendering StartNewGame, w/ keyboard shortcuts 1`] = `
<button
  className="ctrl-btn"
  onClick={[Function]}
>
  Start new
  <br />
  game
  <div
    className="miniKeyHint"
  >
    <div
      className="keyHint miniCtrlHint"
    >
      CTRL
    </div>
    <div
      className="miniPlus"
    >
       + 
    </div>
    <div
      className="keyHint miniEnterHint"
    >
      ENTER
    </div>
  </div>
</button>
`;

exports[`Components: Basic rendering StartNewGame, w/out keyboard shortcuts 1`] = `
<button
  className="ctrl-btn"
  onClick={[Function]}
>
  Start new
  <br />
  game
</button>
`;

exports[`Components: Basic rendering Status, basic render w/ keyboard shortcuts 1`] = `
<Status
  chord="iii"
  chordName="G minor"
  clicksPerRightAnswer={
    Array [
      3.12,
    ]
  }
  displayKeyboardShorcuts={false}
  gameOver={true}
  key_="Eb Major"
  nAnsweredRight={4}
>
  <div
    className="status-box"
  >
    <h3
      className="status-feedback"
    >
      <br />
    </h3>
    <h3
      className="status-secondary"
    >
      4 answered correctly
    </h3>
    <h3
      className="status-secondary"
    >
      3.12 guesses per correct answer
    </h3>
    <h2
      className="status-primary"
    >
      <div>
        Game over! 
        <a
          className="game-link"
          href="javascript:void(0)"
          onClick={[Function]}
        >
          Play again?
        </a>
      </div>
    </h2>
  </div>
</Status>
`;

exports[`Components: Basic rendering Status, basic render, w/out keyboard shortcuts 1`] = `
<Status
  chord="iii"
  chordName="G minor"
  clicksPerRightAnswer={
    Array [
      3,
      12,
    ]
  }
  displayKeyboardShorcuts={false}
  gameOver={true}
  key_="Eb Major"
  nAnsweredRight={4}
>
  <div
    className="status-box"
  >
    <h3
      className="status-feedback"
    >
      <br />
    </h3>
    <h3
      className="status-secondary"
    >
      4 answered correctly
    </h3>
    <h3
      className="status-secondary"
    >
      7.5 guesses per correct answer
    </h3>
    <h2
      className="status-primary"
    >
      <div>
        Game over! 
        <a
          className="game-link"
          href="javascript:void(0)"
          onClick={[Function]}
        >
          Play again?
        </a>
      </div>
    </h2>
  </div>
</Status>
`;
